core/drivers/clk/clk.c

```diff
@@ -0,0 +1,615 @@
+		if (clk_idx == 0)
+			return clk_of_get_from_provider(parent_node,
+							clock_cells,
+							&prop[idx32]);
+
+		clk_idx--;
+		idx += sizeof(phandle) + clock_cells * sizeof(uint32_t);
+	}
+
+	return NULL;
+}
+
+static const struct clk_driver *clk_get_compatible_driver(const char *compat)
+{
+	const struct dt_device_match *dm;
+	const struct dt_driver *drv;
+	const struct clk_driver *clk_drv;
+
+	for_each_dt_driver(drv) {
+		if (drv->type != DT_DRIVER_CLK)
+			continue;
+
+		clk_drv = (const struct clk_driver *)drv->driver;
+		for (dm = drv->match_table; dm; dm++) {
+			if (!dm->compatible)
+				break;
+			if (strcmp(dm->compatible, compat) == 0)
+				return clk_drv;
+		}
+	}
+
+	return NULL;
+}
+
+static void clk_setup_compatible(void *fdt, const char *compatible,
+				 const struct clk_driver *clk_drv);
+
+static void probe_parent_clock(void *fdt, int nodeoffset)
+{
+	const char *compat;
+	int idx, count, len;
+	const struct clk_driver *clk_drv;
+
+	count = fdt_stringlist_count(fdt, nodeoffset, "compatible");
+	if (count < 0)
+		return;
+
+	for (idx = 0; idx < count; idx++) {
+		compat = fdt_stringlist_get(fdt, nodeoffset, "compatible", idx,
+					    &len);
+		if (!compat)
+			return;
+
+		clk_drv = clk_get_compatible_driver(compat);
+		if (clk_drv)
+			clk_setup_compatible(fdt, compat, clk_drv);
+	}
+}
+
+static void parse_clock_property(void *fdt, int nodeoffset)
+{
+	uint32_t phandle;
+	const uint32_t *prop;
+	int len, idx = 0, parent_node, clock_cells;
+
+	prop = fdt_getprop(fdt, nodeoffset, "clocks", &len);
+	if (!prop)
+		return;
+
+	while (idx < len) {
+		phandle = fdt32_to_cpu(prop[idx]);
+
+		parent_node = fdt_node_offset_by_phandle(fdt, phandle);
+		if (parent_node < 0)
+			return;
+
+		probe_parent_clock(fdt, parent_node);
+
+		clock_cells = fdt_clock_cells(fdt, parent_node);
+		idx += sizeof(phandle) + clock_cells * sizeof(uint32_t);
+	}
+}
+
+static void clk_setup_compatible(void *fdt, const char *compatible,
+				 const struct clk_driver *clk_drv)
+{
+	int status;
+	int node = fdt_node_offset_by_compatible(fdt, -1, compatible);
+
+	if (node < 0 || clk_get_provider(node))
+		return;
+
+	while (node >= 0) {
+		status = _fdt_get_status(fdt, node);
+		if (!(status & DT_STATUS_OK_SEC))
+			continue;
+
+		parse_clock_property(fdt, node);
+
+		clk_drv->setup(fdt, node);
+
+		node = fdt_node_offset_by_compatible(fdt, node, compatible);
+	};
+}
+
+static TEE_Result clk_probe(void)
+{
+	const struct dt_device_match *dm;
+	const struct dt_driver *drv;
+	void *fdt = get_embedded_dt();
+	const struct clk_driver *clk_drv;
+
+	IMSG("Probing clocks from device tree\n");
+	if (!fdt)
+		panic();
+
+	for_each_dt_driver(drv) {
+		if (drv->type != DT_DRIVER_CLK)
+			continue;
+
+		clk_drv = (const struct clk_driver *)drv->driver;
+		for (dm = drv->match_table; dm; dm++) {
+			if (!dm->compatible)
+				break;
+
+			clk_setup_compatible(fdt, dm->compatible, clk_drv);
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+service_init(clk_probe);
```

Comment:
Could this be probe as early as possible, like early_init

Reply:
Yes it can be.